File: cs470/proj1/Tennis/TieBreaker.cpp
// Changes made by Zachary Thompson

#include "TieBreaker.hpp"
#include "TieBreakerScore.hpp"
#include "PointScore.hpp"
#include "Player.hpp"

TieBreaker::TieBreaker( Player *p1, Player *p2 ): Competition( p1, p2 ) {}
Score *TieBreaker::play( Player *p ) {
    Score *score = new TieBreakerScore( player1(), player2() );
    Player* server = p;
    int serveCount = 0;

    while ( !score -> haveAWinner() ) {
        PointScore *pointScore = reinterpret_cast<PointScore *>( server->serveAPoint() );
        score->addScore( pointScore->getWinner() );
        delete pointScore;
        server = serveCount++ % 4 <  2 ? Player::otherPlayer(p) : p;
    }

    return score;
}

File: cs470/proj1/ObjectiveCTennis/TennisInObjectiveC/TieBreaker.m
//  Additions made by Zachary Thompson
//  TieBreaker.m
//  TennisInObjectiveC
//
//  Created by student on 2/3/14.
//  Copyright (c) 2014 Ali Kooshesh. All rights reserved.
//

#import "TieBreaker.h"
#import "TieBreakerScore.h"

@implementation TieBreaker

-(Score *) play:(Player *)player
{
    Score *score = [[TieBreakerScore alloc] initWithFirstPlayer:self.player1 secondPlayer:self.player2];
    int serveCount = 0;
    Player* server = player;
    
    while ( ! [score haveAWinner]) {
        Score *pScore = [server serveAPoint];
        [score addScore:[pScore getWinner]];
        pScore = nil;
        server = serveCount++ % 4 < 2 ? [Player otherPlayer:player]: player;
    }
    return score;
}

@end
